Sources/ComposableArchitecture/Effect.swift

```diff
@@ -264,141 +85,380 @@ public struct Effect<Output, Failure: Error>: Publisher {
+  /// in an ``Effect`` so that the reducer, a non-asynchronous context, can process it.
+  ///
+  /// For example, if your environment contains a dependency that exposes an `async` function, you
+  /// can use ``Effect/task(priority:operation:catch:file:fileID:line:)`` to provide an asynchronous
+  /// context for invoking that endpoint:
   ///
   /// ```swift
-  /// case .buttonTapped:
-  ///   return environment.fetchUser(id: 1)
-  ///     .catchToEffect()
-  ///     .map(ProfileAction.userResponse)
+  /// struct FeatureEnvironment {
+  ///   var numberFact: @Sendable (Int) async throws -> String
+  /// }
+  ///
+  /// enum FeatureAction {
+  ///   case factButtonTapped
+  ///   case faceResponse(TaskResult<String>)
+  /// }
+  ///
+  /// let featureReducer = Reducer<State, Action, Environment> { state, action, environment in
+  ///   switch action {
+  ///     case .factButtonTapped:
+  ///       return .task { [number = state.number] in
+  ///         await .factResponse(TaskResult { try await environment.numberFact(number) })
+  ///       }
+  ///
+  ///     case .factResponse(.success(fact)):
+  ///       // do something with fact
+  ///
+  ///     case .factResponse(.failure):
+  ///       // handle error
+  ///
+  ///     ...
+  ///   }
+  /// }
   /// ```
   ///
-  /// - Returns: An effect that wraps `self`.
-  public func catchToEffect() -> Effect<Result<Output, Failure>, Never> {
-    self.map(Result.success)
-      .catch { Just(.failure($0)) }
-      .eraseToEffect()
+  /// The above code sample makes use of ``TaskResult`` in order to automatically bundle the success
+  /// or failure of the `numberFact` endpoint into a single type that can be sent in an action.
+  ///
+  /// The closure provided to ``Effect/task(priority:operation:catch:file:fileID:line:)`` is allowed
+  /// to throw, but any non-cancellation errors thrown will cause a runtime warning when run in the
+  /// simulator or on a device, and will cause a test failure in tests.
+  ///
+  /// - Parameters:
+  ///   - priority: Priority of the underlying task. If `nil`, the priority will come from
+  ///     `Task.currentPriority`.
+  ///   - operation: The operation to execute.
+  ///   - catch: An error handler, invoked if the operation throws an error other than
+  ///     `CancellationError`.
+  /// - Returns: An effect wrapping the given asynchronous work.
+  public static func task(
+    priority: TaskPriority? = nil,
+    operation: @escaping @Sendable () async throws -> Output,
+    catch handler: (@Sendable (Error) async -> Output)? = nil,
+    file: StaticString = #file,
+    fileID: StaticString = #fileID,
+    line: UInt = #line
+  ) -> Self {
+    Deferred<Publishers.HandleEvents<PassthroughSubject<Output, Failure>>> {
+      let subject = PassthroughSubject<Output, Failure>()
+      let task = Task(priority: priority) { @MainActor in
+        await withTaskCancellationHandler {
+          if Thread.isMainThread {
```

Comment:
I'm not sure, but Thread.isMainThread may return invalid results (when working with Dispatch), this's why ReactiveSwift framework introduced UIScheduler, see https://github.com/ReactiveCocoa/ReactiveSwift/blob/master/Sources/Scheduler.swift#L92

Reply:
@larryonoff Thanks for going through the PR with a fine toothed comb and uncovering this. 🙂

We are actually going to remove the cancellation handler from `Effect.task` and `Effect.run` for right now. We only added it to support doing:

```swift
withUnsafeCurrentTask { $0?.cancel() }
```

...but that actually kind of behaves weirdly and we're not sure if it's a Swift concurrency bug or what.

Also, we're not sure handling cancellation in this way is correct, otherwise _every_ unstructured task ever created should technically start with a `withTaskCancellationHandler`, which would be weird.

We will keep thinking about it and see if we should bring it back later.